File: webserver/response.py
from . import util
from . import status

class Response:
    def __init__(self, protocol="HTTP/1.1",statuscode=200,headers={"server": "spellsoftruth"},body={"length":0,"content":None}):
        self.protocol = protocol
        self.statuscode = statuscode
        self.headers = headers
        self.body = body

    def addheader(self,name, value):
        self.headers[name] = value

    def setcookie(self,name,value,attributes={}):
        buf = f"{util.urlencode(name)}={util.urlencode(value)}"
        for k,v in attributes.items():
            buf += f" {k}={v}"
        self.addheader("Set-Cookie", str(buf))

    def toheaderstring(self):
        sb = f"{self.protocol} {self.statuscode} {status.getstatus(self.statuscode)}\r\n"
        for key,value in self.headers.items():
            # print(f"toheaderstring key:{key} value:{value}")
            if isinstance(value, list):
                for v in value:
                    sb += f"{key}: {v}\r\n"
            else:
                sb += f"{key}: {value}\r\n"
        sb += "\r\n"
        print(f"response toheaderstring:{sb}")
        return sb

    @staticmethod
    def errorresponse(statuscode, text):
        response = Response()
        response.statuscode = int(statuscode)
        response.headers["content-type"] = "text/plain; charset=utf-8"
        response.body['content'] = text
        response.body['length'] = len(text)
        return response

File: goodjava/src/goodjava/webserver/Response.java
package goodjava.webserver;

import java.io.InputStream;
import java.io.PrintWriter;
import java.util.Map;
import java.util.LinkedHashMap;
import java.util.Collections;
import java.util.List;


public class Response {
	public final String protocol = "HTTP/1.1";
	public volatile Status status = Status.OK;
	public final Map<String,Object> headers = Collections.synchronizedMap(new LinkedHashMap<String,Object>());
	{
		headers.put("server","Goodjava");
	}
	private static final Body empty = new Body(0,new InputStream(){
		public int read() { return -1; }
	});
	public volatile Body body = empty;

	public static class Body {
		public final long length;
		public final InputStream content;
	
		public Body(long length,InputStream content) {
			this.length = length;
			this.content = content;
		}
	}


	public void addHeader(String name,String value) {
		Util.add(headers,name,value);
	}

	public void setCookie(String name,String value,Map<String,String> attributes) {
		StringBuilder buf = new StringBuilder();
		buf.append( Util.urlEncode(name) );
		buf.append( '=' );
		buf.append( Util.urlEncode(value) );
		for( Map.Entry<String,String> entry : attributes.entrySet() ) {
			buf.append( "; " );
			buf.append( entry.getKey() );
			buf.append( '=' );
			buf.append( entry.getValue() );
		}
		addHeader( "Set-Cookie", buf.toString() );
	}


	public String toHeaderString() {
		StringBuilder sb = new StringBuilder();
		sb.append( protocol )
			.append( ' ' ).append( status.code )
			.append( ' ' ).append( status.reason )
			.append( "\r\n" )
		;
		for( Map.Entry<String,Object> entry : headers.entrySet() ) {
			String name = entry.getKey();
			Object value = entry.getValue();
			if( value instanceof List ) {
				for( Object v : (List)value ) {
					sb.append( name ).append( ": " ).append( v ).append( "\r\n" );
				}
			} else {
				sb.append( name ).append( ": " ).append( value ).append( "\r\n" );
			}
		}
		sb.append( "\r\n" );
		return sb.toString();
	}


	public static Response errorResponse(Status status,String text) {
		Response response = new Response();
		response.status = status;
		response.headers.put( "content-type", "text/plain; charset=utf-8" );
		PrintWriter writer = new PrintWriter( new ResponseOutputStream(response) );
		writer.write( text );
		writer.close();
		return response;
	}
}
